src/options.js

```diff
@@ -0,0 +1,416 @@
+ * This is in addition to the calculation derived from the <i>request_min_throughput</i> parameter.
+ * </li>
+ * <li> request_min_throughput: <i>integer</i> - Minimum expected throughput in bytes per second for HTTP requests, set to 100 KiB/s seconds by default. <br>
+ * If the throughput is lower than this value, the connection will time out. This is used to calculate an additional
+ * timeout on top of <i>request_timeout</i>. This is useful for large requests. You can set this value to 0 to disable this logic.
+ * </li>
+ * <li> retry_timeout: <i>integer</i> - The time in milliseconds to continue retrying after a failed HTTP request, set to 10 seconds by default. <br>
+ * The interval between retries is an exponential backoff starting at 10ms and doubling after each failed attempt up to a maximum of 1 second.
+ * </li>
+ * </ul>
+ * <br>
+ * Other options
+ * <ul>
+ * <li> max_name_len: <i>integer</i> - The maximum length of a table or column name, the Sender defaults this parameter to 127. <br>
+ * Recommended to use the same setting as the server, which also uses 127 by default.
+ * </li>
+ * <li> copy_buffer: <i>enum, accepted values: on, off</i> - By default, the Sender creates a new buffer for every flush() call,
+ * and the data to be sent to the server is copied into this new buffer.
+ * Setting the flag to <i>off</i> results in reusing the same buffer instance for each flush() call. <br>
+ * Use this flag only if calls to the client are serialised.
+ * </li>
+ * </ul>
+ */
+class SenderOptions {
+
+    protocol;
+    addr;
+    host; // derived from addr
+    port; // derived from addr
+
+    // replaces `auth` and `jwk` options
+    username;
+    password;
+    token;
+    token_x; // allowed, but ignored
+    token_y; // allowed, but ignored
+
+    auto_flush;
+    auto_flush_rows;
+    auto_flush_interval;
+
+    // replaces `copyBuffer` option
+    copy_buffer;
+
+    request_min_throughput;
+    request_timeout;
+    retry_timeout;
+
+    // replaces `bufferSize` option
+    init_buf_size;
+    max_buf_size;
+
+    tls_verify;
+    tls_ca;
+    tls_roots;          // not supported
+    tls_roots_password; // not supported
+
+    max_name_len;
+
+    log;
+    agent;
+
+    /**
+     * Creates a Sender options object by parsing the provided configuration string.
+     *
+     * @param {string} configurationString - Configuration string. <br>
+     * @param {object} extraOptions - Optional extra configuration. <br>
+     * 'log' is a logging function used by the <a href="Sender.html">Sender</a>.
+     * Prototype: <i>(level: 'error'|'warn'|'info'|'debug', message: string) => void</i>. <br>
+     * 'agent' is a custom http/https agent used by the <a href="Sender.html">Sender</a> when http/https transport is used. <br>
+     * A <i>http.Agent</i> or <i>https.Agent</i> object is expected.
+     */
+    constructor(configurationString, extraOptions = undefined) {
```

Comment:
nit: in JS any omitted argument is `undefined`:
```suggestion
    constructor(configurationString, extraOptions) {
```

Reply:
just realised why i have these.
it makes the parameter optional in the typescript declaration files.
```
constructor(configurationString: string, extraOptions?: object);
```

added it back.

Reply:
I see. Let's keep them then.